src/cljs/athens/views/find_in_page.cljs

```diff
@@ -0,0 +1,243 @@
+
+
+(defn find-in-page!
+  "Uses webContents findInPage(browser based highlighting.
+   Handles back and forward search"
+  ([] (find-in-page! (:text @(subscribe [:find-in-page-info]))))
+  ([text] (find-in-page! text false))
+  ([text-to-search back?]
+   (when (some-> text-to-search seq str/join)
+     (let [opts (cond-> {:forward  true
+                         :findNext false}
+
+                  (= text-to-search
+                     (:text @(subscribe [:find-in-page-info])))
+                  (merge {:findNext true})
+
+                  back? (merge {:forward false})
+                  true clj->js)]
+
+       (.. (window-id->window @!main-window-id)
+           -webContents
+           (findInPage text-to-search opts))))))
+
+
+;; --------------------------------------------------------------------
+;; ---- subs---
+
+
+(rf/reg-sub
+  :find-in-page-info
+  (fn [db]
+    (:find-in-page-info db)))
+
+
+;; --------------------------------------------------------------------
+;; ---- evts ---
+
+(rf/reg-fx :find-in-page find-in-page!)
+
+
+(rf/reg-event-fx
+  :set-find-in-page-text
+  (fn [{:keys [db]} [_ text]]
+    (merge
+      {:db           (assoc-in db [:find-in-page-info :text] text)
+       :find-in-page text}
+      (when (str/blank? text)
+        {:clear-current-selection _}))))
+
+
+(rf/reg-event-fx
+  :next-result
+  (fn [_ _] (find-in-page!)))
+
+
+(rf/reg-fx
+  :prev-result
+  (fn []
+    (find-in-page!
+      (:text @(subscribe [:find-in-page-info])) true)))
+
+
+;; --------------------------------------------------------------------
+;; ---- messages from find window ---
+
+
+(declare stop-find-in-page!)
+
+
+;; listens to messages from find-window
+(defonce __ipc-listener-main__
+  (.. electron -remote -ipcMain
+      (on "find-window->parent"
+          (fn [_ msg]
+            (let [[event & args] (js->clj msg)]
+              (case (keyword event)
+                :text (dispatch [:set-find-in-page-text (first args)])
+                :close (stop-find-in-page!)
+                :next (find-in-page!)
+
+                :prev
+                (find-in-page!
+                  (:text @(subscribe [:find-in-page-info]))
+                  true)))))))
+
+
+;; --------------------------------------------------------------------
+;; ---- life cycle ---
+
+
+(defn send-msg-to-find-window!
+  [& msg]
+  (.. electron -ipcRenderer
+      (send "parent->find-window"
+            (-> msg vec clj->js))))
+
+
+(defn get-find-win-pos
+  ([] (get-find-win-pos
+        (window-id->window @!main-window-id)))
+  ([win]
+   (let [[x y] (-> win (. getPosition) js->clj)
+         [_ y-win-size] (-> win (. getSize) js->clj)
+         [_ y-content-size] (-> win (. getContentSize) js->clj)]
+     [(+ x 20)
+      ;; + y 45(below toolbar) 25(title bar)
+      (+ y 45 (when-not (. win isFullScreen)
+                (- y-win-size y-content-size)))])))
+
+
+(defn init!
+  []
+  (let [browser-win (.. electron -remote -BrowserWindow)
```

Comment:
Are there any solutions that don't create a new BrowserWindow?

I get that this solution let's us use native `electron/webContents` lib, but it introduces some cruft: another file for keybinding listeners, another html file with font imports, IPC usage, another shadow-cljs build target (see separate comment on this).

A separate window may also introduce more surface area for OS-specific edge cases.

For example, I use a tiling manager.

![image](https://user-images.githubusercontent.com/8952138/119848483-b4211f80-bec0-11eb-8666-011c90b9aafb.png)

![image](https://user-images.githubusercontent.com/8952138/119848725-e894db80-bec0-11eb-9d98-94eb0eb8b5aa.png)

Reply:
Tiling manager is able to control the find in page window? Is it not changing with position of the parent as mac does? is it not changing when window resized?


Webcontents api doesnt have the exclusion of what not to highlight so each highlight will also highlight the text field. You would have to deal with refocusing logic on input and could have more issues.

It's basically mimics chrome window

Reply:
Need more info on this.

Reply:
Is this another valid solution? https://discord.com/channels/708122962422792194/708124156113321985/847631235320905738

Reply:
> Is this another valid solution? https://github.com/glafche/electron-find-in-page

There were already other deps that did this? This is not really different.
https://github.com/rhysd/electron-in-page-search did similar(more robust)
All these are literal wrappers on webContents API and provide a window(that's it). Shallow modules and not much value. Constrained API for the wrapper itself. Why not depend on the robust browser API of webContents and avoid these unnecessary deps?

> What is wrong with highlight.js?

it modifies elements in place. Will have unintended conseq on rendering logic.
Way too slow if there are too many matches.

> Sometimes the findWindow does not update position with parent.

Maybe old code, if reloaded main window loses findWindow that's already open. 0e3bfa8 commit prevents reloads in such cases. If it's happening after than commit, need more info.

> Tiling manager

I don't see it happening in my mac. I don't have access to a ubuntu machine. Might need some help here. 
I doubt adding an npm dep will fix this because it's just a window at the end.

Reply:
> Creating a separate window is not mimicking Chrome in a pretty big way?

This is probably the best electron offers. Not sure if there are any other ways for "attached window" to be present in electron. Functionality is the one that's getting mimicked. What do you think is limiting in terms of functionality?

Reply:
Is it possible to back up and understand the tradeoffs between using webContents API versus creating an element within the original window?

I think this is a good architectural question that merits a design document. There are both engineering and product considerations between the two approaches (maybe there are even more than 2 approaches).

Is it possible to back up and understand the tradeoffs between using webContents API versus creating an element within the original window?

I think this is a good architectural question that merits a design document. There are both engineering and product considerations between the two approaches (maybe there are even more than 2 approaches).

Obsidian seems to create DOM elements directly with their find-in-page. They are probably not the only ones doing this so maybe we can learn from them.

Reply:
Is it possible to back up and understand the tradeoffs between using webContents API versus creating an element within the original window?

I think this is a good architectural question that merits a design document. There are both engineering and product considerations between the two approaches (maybe there are even more than 2 approaches).

Is it possible to back up and understand the tradeoffs between using webContents API versus creating an element within the original window?

I think this is a good architectural question that merits a design document. There are both engineering and product considerations between the two approaches (maybe there are even more than 2 approaches).

Obsidian seems to create DOM elements directly with their find-in-page. They are probably not the only ones doing this so maybe we can learn from them.

Reply:
41ab9e0

Reply:
Is it possible to back up and understand the tradeoffs between using webContents API versus creating an element within the original window?

I think this is a good architectural question that merits a design document. There are both engineering and product considerations between the two approaches (maybe there are even more than 2 approaches).

Is it possible to back up and understand the tradeoffs between using webContents API versus creating an element within the original window?

I think this is a good architectural question that merits a design document. There are both engineering and product considerations between the two approaches (maybe there are even more than 2 approaches).

Obsidian seems to create DOM elements directly with their find-in-page. They are probably not the only ones doing this so maybe we can learn from them.

Reply:
tangjeff0-bot?

Reply:
lol bad internet

Reply:
Let me know if the reasoning in adr sounds ok. That's one blocker along with the tiling issue. Let's debug that if possible(I don't have access to an ubuntu machine)

Reply:
Tiling is reproducible in mac. Explored modal views(basically there's no good customization to it in electron -- can't even change the backdrop or position).

added code to force find window back to the position but, that doesn't stop tiling manager to create space.
https://www.loom.com/share/cad9928b62394f1bb713f1714a429392 

I think this is the best possible with electron + window. It will server most users(not sure how many use a tiling manager). Others alternatives are highlighted in adr.